src/crypto_provider.rs

```diff
@@ -0,0 +1,273 @@
+use libc::size_t;
+use std::slice;
+use std::sync::Arc;
+
+use rustls::crypto::ring;
+use rustls::crypto::CryptoProvider;
+use rustls::SupportedCipherSuite;
+
+use crate::cipher::rustls_supported_ciphersuite;
+use crate::{
+    arc_castable, box_castable, ffi_panic_boundary, free_arc, free_box, rustls_result,
+    set_arc_mut_ptr, set_boxed_mut_ptr, to_arc_const_ptr, to_boxed_mut_ptr, try_clone_arc,
+    try_mut_from_ptr, try_mut_from_ptr_ptr, try_ref_from_ptr, try_ref_from_ptr_ptr, try_slice,
+    try_take,
+};
+
+box_castable! {
+    /// A `rustls_crypto_provider` builder.
+    pub struct rustls_crypto_provider_builder(Option<CryptoProviderBuilder>);
+}
+
+/// A builder for customizing a `CryptoProvider`. Can be used to install a process-wide default.
```

Comment:
Another way to improve on this might be to rename `_new` to `_new_from_default`. Having the name end in just `_new` makes it seem like it's the default thing to use first, when really they have approximately equal relevance to most people.

Reply:
> So you can't, for instance, for rustls_provider_builder_new, then call rustls_crypto_provider_builder_build_as_default on the result, because either the first call or the second call will error.

That's correct. 

> I think the intent is that you do _new_with_base, followed by _build_as_default to set the process-wide default, and then for subsequent tweaks (e.g. on a per-connection basis), you can do _new. Is that right? 

Yup! You got it. The only extra nuance is that if only one provider is enabled, the default is implicit and nothing needs to explicitly set it up-front. I'll touch on this more in a separate reply to your question about adding CI coverage for builds with both ring/aws-lc-rs enabled (spoiler: I've tried to avoid supporting that for rustls-ffi).

>  If so I'd love to see it explained in a comment for the type, and maybe for each of the functions. 

I've taken a crack at expressing the usual "workflows" for the builders in additions to the comments. PTAL!

>  Another way to improve on this might be to rename _new to _new_from_default

That's a good idea. I implemented this suggested rename.